libraries/AP_Math/quaternion.cpp

```diff
@@ -22,17 +22,43 @@
 #include <AP_InternalError/AP_InternalError.h>
 
 // return the rotation matrix equivalent for this quaternion
-void Quaternion::rotation_matrix(Matrix3f &m) const
-{
-    const float q3q3 = q3 * q3;
-    const float q3q4 = q3 * q4;
-    const float q2q2 = q2 * q2;
-    const float q2q3 = q2 * q3;
-    const float q2q4 = q2 * q4;
-    const float q1q2 = q1 * q2;
-    const float q1q3 = q1 * q3;
-    const float q1q4 = q1 * q4;
-    const float q4q4 = q4 * q4;
+template <typename T>
+void QuaternionT<T>::rotation_matrix(Matrix3d &m) const
+{
+    const T q3q3 = q3 * q3;
+    const T q3q4 = q3 * q4;
+    const T q2q2 = q2 * q2;
+    const T q2q3 = q2 * q3;
+    const T q2q4 = q2 * q4;
+    const T q1q2 = q1 * q2;
+    const T q1q3 = q1 * q3;
+    const T q1q4 = q1 * q4;
+    const T q4q4 = q4 * q4;
+
+    m.a.x = 1.0f-2.0f*(q3q3 + q4q4);
```

Comment:
This uses a different form to the numerator for line 90 in rotation_matrix_norm. Is there any reason for this?

Reply:
it hasn't changed in this PR, just a new version of rotation_matrix() that takes a double as input is added, I don't know why rotation_matrix_norm has the form that it has
actually, rotation_matrix_norm() is totally unused, I'll just remove it